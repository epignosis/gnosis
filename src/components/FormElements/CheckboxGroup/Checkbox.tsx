import React, { forwardRef, ForwardRefRenderFunction } from "react";
import { SerializedStyles } from "@emotion/react";
import { checkboxContainer } from "./styles";
import { ExtendableProps } from "types/common";

export type CheckboxOption = {
  value: string;
  label: string;
  name: string;
  disabled?: boolean;
};

export type CheckboxSize = "md" | "lg";

export type CheckboxProps = ExtendableProps<
  Omit<React.InputHTMLAttributes<HTMLInputElement>, "type">,
  CheckboxOption & {
    id: string;
    size?: CheckboxSize;
    inline?: boolean;
    containerAttrs?: React.HTMLAttributes<HTMLDivElement>;
  }
>;

const Checkbox: ForwardRefRenderFunction<HTMLDivElement, CheckboxProps> = (props, ref) => {
  const { id, label, size = "md", inline = false, containerAttrs, ...rest } = props;

  return (
    <div
      ref={ref}
      css={(theme): SerializedStyles => checkboxContainer(theme, { size, inline })}
      {...containerAttrs}
    >
      <input id={id} type="checkbox" {...rest} />
      <label htmlFor={id}>
        <span className="shadow-element" tabIndex={-1} aria-hidden="true" />
        {label}
      </label>
    </div>
  );
};

export default forwardRef(Checkbox);
