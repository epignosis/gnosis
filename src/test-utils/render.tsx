import React, { FC, ReactElement, ComponentType } from "react";
import { render, RenderResult, RenderOptions } from "@testing-library/react";
import ThemeProvider from "../components/ThemeProvider/ThemeProvider";

const AllTheProviders: FC = ({ children }) => <ThemeProvider>{children}</ThemeProvider>;

const customRender = (ui: ReactElement, options: RenderOptions): RenderResult =>
  render(ui, { wrapper: AllTheProviders as ComponentType, ...options });

export * from "@testing-library/react";

export { customRender as render };
