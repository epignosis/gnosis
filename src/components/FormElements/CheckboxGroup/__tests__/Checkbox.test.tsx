import React from "react";
import userEvent from "@testing-library/user-event";
import faker from "faker";
import Checkbox from "../Checkbox";
import { render, screen } from "@test-utils/render";

describe("<Checkbox />", () => {
  it("renders correctly", () => {
    const labelTxt = faker.lorem.word();
    const id = faker.random.alphaNumeric();
    const name = faker.random.alphaNumeric();
    render(<Checkbox id={id} label={labelTxt} name={name} value="testValue" />);

    const label = screen.getByText(labelTxt);
    const input = screen.getByLabelText(labelTxt);

    expect(label).toHaveTextContent(labelTxt);
    expect(input).toHaveAttribute("name", name);
    expect(input).toHaveAttribute("id", id);
    expect(input).not.toBeChecked();
  });

  it("changes from un-checked to checked on click", () => {
    const labelTxt = faker.lorem.word();
    const id = faker.random.alphaNumeric();
    const name = faker.random.alphaNumeric();
    render(
      <Checkbox id={id} label={labelTxt} name={name} value="testValue" onChange={jest.fn()} />,
    );

    const label = screen.getByText(labelTxt);
    const input = screen.getByLabelText(labelTxt);

    expect(input).not.toBeChecked();

    userEvent.click(input);

    expect(input).toBeChecked();

    userEvent.click(label);

    expect(input).not.toBeChecked();
  });

  it("is disabled", () => {
    const labelTxt = faker.lorem.word();
    const id = faker.random.alphaNumeric();
    const name = faker.random.alphaNumeric();
    render(<Checkbox id={id} label={labelTxt} name={name} value="testValue" disabled />);

    const label = screen.getByText(labelTxt);
    const input = screen.getByLabelText(labelTxt);

    expect(input).not.toBeChecked();

    userEvent.click(label);

    expect(input).not.toBeChecked();
  });

  it("matches snapshot with `md` size", () => {
    const { container } = render(
      <Checkbox
        id="my-id"
        className="checkbox"
        label="Test label"
        name="testName"
        value="testValue"
        containerAttrs={{
          id: "my-container-id",
          className: "container-class",
        }}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with `lg` size", () => {
    const { container } = render(
      <Checkbox size="lg" id="testId" label="Test label" name="testName" value="testValue" />,
    );

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with `inline` = true", () => {
    const { container } = render(
      <Checkbox id="testId" label="Test label" name="testName" value="testValue" inline />,
    );

    expect(container).toMatchSnapshot();
  });
});
