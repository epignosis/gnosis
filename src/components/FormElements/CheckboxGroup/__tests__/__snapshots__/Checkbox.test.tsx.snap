// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> matches snapshot with \`inline\` = true 1`] = `
<div>
  <div
    class="css-6emo4e-checkboxContainer-checkboxContainer"
  >
    <input
      id="testId"
      name="testName"
      type="checkbox"
      value="testValue"
    />
    <label
      for="testId"
    >
      <span
        aria-hidden="true"
        class="shadow-element"
        tabindex="-1"
      />
      Test label
    </label>
  </div>
</div>
`;

exports[`<Checkbox /> matches snapshot with \`lg\` size 1`] = `
<div>
  <div
    class="css-1p9f0hm-checkboxContainer-checkboxContainer"
  >
    <input
      id="testId"
      name="testName"
      type="checkbox"
      value="testValue"
    />
    <label
      for="testId"
    >
      <span
        aria-hidden="true"
        class="shadow-element"
        tabindex="-1"
      />
      Test label
    </label>
  </div>
</div>
`;

exports[`<Checkbox /> matches snapshot with \`md\` size 1`] = `
<div>
  <div
    class="css-1j4ss74-checkboxContainer-checkboxContainer"
  >
    <input
      id="testId"
      name="testName"
      type="checkbox"
      value="testValue"
    />
    <label
      for="testId"
    >
      <span
        aria-hidden="true"
        class="shadow-element"
        tabindex="-1"
      />
      Test label
    </label>
  </div>
</div>
`;
